### PR TITLE
[Natalia Falinska] UIT 34 recently viewed

### PR DESCRIPTION
**Pull request by NFalinska**

### CODE REVIEW

**Comment by NFalinska** [comment]
https://issues.griddynamics.net/secure/RapidBoard.jspa?rapidView=298&view=detail&selectedIssue=UIT-34

File: src/app/app.component.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import { Component } from '@angular/core';
 import { Router, NavigationEnd } from '@angular/router';
 import { environment } from 'src/environments/environment';
+// import { RecentlyViewedService } from './shared/components/recently-viewed/recently-viewed.service';
```
Comment (dverbovyi) [inline src/app/app.component.ts]: wtf?

File: src/app/app.component.ts [hunk 2]
```diff
@@ -17,6 +18,8 @@ export class AppComponent {
         this.isEnvironmentRoute = true;
       }
     });
+
+    // recentlyViewedService.loadRouting();
```
Comment (dverbovyi) [inline src/app/app.component.ts]: wtf?

File: src/app/product-details-page/components/product-details-page/product-details-page.component.html [hunk 3]
```diff
@@ -1 +1,3 @@
 <app-product-description [product]="productInfo"></app-product-description>
+
+<!-- <p>{{stored.id}}</p> -->
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.html]: ?

File: src/app/home-page/components/home-page/home-page.component.html [hunk 4]
```diff
@@ -1,29 +1,23 @@
 <app-slideshow></app-slideshow>
 <app-heading [alertText]="'new'" [headingText]="'Arrivals'"></app-heading>
-<app-carousel>
+<app-carousel [productsLength]="products.length">
```
Comment (dverbovyi) [inline src/app/home-page/components/home-page/home-page.component.html]: revert this change

File: src/app/home-page/components/home-page/home-page.component.html [hunk 5]
```diff
@@ -1,29 +1,23 @@
...
 <app-heading id="wish-list" *ngIf="wishListFilter.length" [alertText]="'wish'" [headingText]="'list'"></app-heading>
 
-<div class="main-container">
-  <div class="product-item-short-wrapper">
-    <ng-container *ngFor="let product of wishListFilter; let i = index">
-      <app-product-item-short  *ngIf="i < 3" [product]="product"></app-product-item-short>
-    </ng-container>
-  </div>
-</div>
+<app-carousel *ngIf="viewedItems" [productsLength]="viewedItems.length">
```
Comment (dverbovyi) [inline src/app/home-page/components/home-page/home-page.component.html]: create recently viewed component

File: src/app/home-page/components/home-page/home-page.component.ts [hunk 6]
```diff
@@ -10,6 +10,8 @@ import { IProduct, ProductSize, ProductAvailabilityState } from 'src/app/shared/
 
 
 export class HomePageComponent {
+  viewedItems = JSON.parse(localStorage.getItem('recentlyViewed'));
```
Comment (dverbovyi) [inline src/app/home-page/components/home-page/home-page.component.ts]: move to recently viewed component

File: src/app/product-details-page/components/product-details-page/product-details-page.component.ts [hunk 7]
```diff
@@ -24,4 +23,11 @@ export class ProductDetailsPageComponent {
     title: 'Half Jacket + Skiny Trousers + Boot leather',
     description: 'Lorem Lorem Lorem',
   };
+
+  ngOnInit() {
+    const stored = JSON.parse(localStorage.getItem('recentlyViewed')) || [];
+    const newProd = this.productInfo;
+    stored.unshift(newProd);
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.ts]: only id has to be stored

**Review by dverbovyi — CHANGES_REQUESTED** [review]

File: src/app/home-page/components/home-page/home-page.component.html (hunk 4 above)
Comment (dverbovyi) [reply src/app/home-page/components/home-page/home-page.component.html]: ping

File: src/app/product-details-page/components/product-details-page/product-details-page.component.ts [hunk 8]
```diff
@@ -24,4 +24,28 @@ export class ProductDetailsPageComponent {
...
+
+  constructor(
+    private activatedRoute: ActivatedRoute
+  ) { }
+
+  ngOnInit() {
+    const stored = JSON.parse(localStorage.getItem('recentlyViewed')) || [];
+    const newProdId = this.activatedRoute.snapshot.params['id'];
+    if (stored.length) {
+      let isInArray = 0;
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.ts]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

File: src/app/shared/components/carousel/carousel.component.scss [hunk 9]
```diff
@@ -61,7 +61,7 @@
 .slides-container {
   display: flex;
   transition: 0.8s;
-  justify-content: space-between;
+  // justify-content: space-between;
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.scss]: ping

File: src/app/shared/components/carousel/carousel.component.ts [hunk 10]
```diff
@@ -17,6 +17,7 @@ export class CarouselComponent implements OnInit, AfterViewInit, OnDestroy {
   private counterScrolledItems: number;
 
   @ViewChild('slidesContainer') slidesContainer: ElementRef;
+  @Input() productsLength;
```
Comment (dverbovyi) [inline src/app/shared/components/carousel/carousel.component.ts]: revert all changes in carousel

File: src/app/shared/components/recently-viewed/recently-viewed.component.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+import { ProductAvailabilityState, ProductSize, IProduct } from '../../interfaces/product';
+
+@Component({
+  selector: 'app-recently-viewed',
+  templateUrl: './recently-viewed.component.html',
+  styleUrls: ['./recently-viewed.component.scss']
+})
+export class RecentlyViewedComponent implements OnInit {
+
+  viewedItems: IProduct [] = [{
```
Comment (dverbovyi) [inline src/app/shared/components/recently-viewed/recently-viewed.component.ts]: pass products as Input property

File: src/app/shared/components/recently-viewed/recently-viewed.component.html [hunk 12]
```diff
@@ -0,0 +1,5 @@
+<app-carousel *ngIf="viewedItems.length" [productsLength]="viewedItems.length">
```
Comment (dverbovyi) [inline src/app/shared/components/recently-viewed/recently-viewed.component.html]: where is app-heading component with an appropriate title?

File: src/app/home-page/components/home-page/home-page.component.html [hunk 13]
```diff
@@ -1,29 +1,19 @@
...
 <app-heading id="wish-list" *ngIf="wishListFilter.length" [alertText]="'wish'" [headingText]="'list'"></app-heading>
 
-<div class="main-container">
-  <div class="product-item-short-wrapper">
-    <ng-container *ngFor="let product of wishListFilter; let i = index">
-      <app-product-item-short  *ngIf="i < 3" [product]="product"></app-product-item-short>
-    </ng-container>
-  </div>
-</div>
+<app-recently-viewed></app-recently-viewed>
```
Comment (dverbovyi) [inline src/app/home-page/components/home-page/home-page.component.html]: <app-recently-viewed *ngIf="products.length" [products]="products"></app-recently-viewed>

**Review by dverbovyi — CHANGES_REQUESTED** [review]

File: src/app/product-details-page/components/product-details-page/product-details-page.component.ts [hunk 14]
```diff
@@ -13,19 +13,37 @@ import { IProduct } from 'src/app/shared/interfaces/product';
   styleUrls: ['./product-details-page.component.scss']
 })
 export class ProductDetailsPageComponent implements OnInit, OnDestroy {
-
   public product: IProduct;
   private productSubscription;
-
+  productInfo = {
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.ts]: remove this mocked object. it's not using anymore

File: src/app/product-details-page/components/product-details-page/product-details-page.component.ts [hunk 15]
```diff
@@ -13,19 +13,37 @@ import { IProduct } from 'src/app/shared/interfaces/product';
...
+    sizes: [ProductSize.S, ProductSize.M, ProductSize.L, ProductSize.XL],
+    addedToCart: false,
+    addedToWishList: true,
+    title: 'Half Jacket + Skiny Trousers + Boot leather',
+    description: 'Lorem Lorem Lorem',
+  };
   constructor(private route: ActivatedRoute, private productService: ProductDetailsPageService) { }
-
   ngOnInit() {
+    const stored = JSON.parse(localStorage.getItem('recentlyViewed')) || [];
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.ts]: rename to 

`const recentlyViewedIds = JSON.parse(localStorage.getItem('recentlyViewedIds'))`

File: src/app/shared/components/recently-viewed/recently-viewed.component.html [hunk 16]
```diff
@@ -0,0 +1,6 @@
+<app-heading *ngIf="products.length" [alertText]="'recently viewed'" [headingText]="'list'"></app-heading>
```
Comment (dverbovyi) [inline src/app/shared/components/recently-viewed/recently-viewed.component.html]: recently viewed component should not render at all if products are absent. So please remove this check as you have it in the parent component

File: src/app/shared/components/recently-viewed/recently-viewed.component.html [hunk 17]
```diff
@@ -0,0 +1,6 @@
...
+<app-carousel *ngIf="products.length">
```
Comment (dverbovyi) [inline src/app/shared/components/recently-viewed/recently-viewed.component.html]: the same as above

File: src/app/product-details-page/components/product-details-page/product-details-page.component.ts [hunk 18]
```diff
@@ -13,19 +13,37 @@ import { IProduct } from 'src/app/shared/interfaces/product';
...
+    const newProdId = this.route.snapshot.params['id'];
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.ts]: move all this logic related to recently viewed into a separate function called markAsRecentlyViewed

File: src/app/shared/components/recently-viewed/recently-viewed.component.ts [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+ @Input() products: IProduct [] = [];
+
+  constructor() { }
+
+  ngOnInit() {
+    const recentlyViewIds = JSON.parse(localStorage.getItem('recentlyViewed'));
+    if (recentlyViewIds) {
+      this.products = this.products.filter( el => recentlyViewIds.some(e => e === el.id ));
+    } else {
+      this.products = [];
```
Comment (dverbovyi) [inline src/app/shared/components/recently-viewed/recently-viewed.component.ts]: else isn't needed here

**Review by dverbovyi — CHANGES_REQUESTED** [review]
@NFalinska good job, please fix minor comments

**Review by dverbovyi — APPROVED** [review]

**Pull request merged**